File: tools/skyepg/huffman.py
#!/usr/bin/env python3
"""Sky/OpenTV EPG Huffman codec, built from the published Sky UK dictionary.

WHY A CODEC AND NOT JUST A DECODER. Every public implementation of this format READS the stream;
we need to WRITE one, which means encoding. An encoder has a property a decoder does not: it can be
validated against the decoder it is meant to feed. So this file carries both halves and a
round-trip check, and nothing here is trusted until a string survives encode -> decode unchanged.

THE DICTIONARY FORMAT, read off openTVtoXML's huffman_read_dictionary() rather than guessed:
each line is `<value>=<bits>`, parsed by three sscanf attempts in order --
    "%c=%[^\\n]"        a single character (so a line starting with a space maps the SPACE char)
    "%[^=]=%[^\\n]"     a multi-character phrase
    "=%[^\\n]"          an EMPTY value: a node that emits nothing and resets to the root
That ordering matters: a naive split on '=' mis-parses both the space line and any phrase
containing '='.

THE BIT PACKING IS THE PART THAT WILL CATCH PEOPLE. The decoder starts the FIRST byte at mask 0x20
and every later byte at 0x80 -- so byte 0 contributes only its low SIX bits and its top two are
skipped entirely. That is not a typo in the reference implementation; it is the format.
"""
import sys
from pathlib import Path

DEFAULT_DICT = Path(__file__).resolve().parent / 'skyuk.dict'


def read_dictionary(path):
    """value -> bits, mirroring huffman_read_dictionary()'s three-way parse."""
    table, terminator = {}, None
    for raw in Path(path).read_text(encoding='latin-1').splitlines():
        if not raw or '=' not in raw:
            continue
        value, _, code = raw.partition('=')
        if not set(code) <= {'0', '1'} or not code:
            continue
        if value == '':
            terminator = code            # the empty-value node: emits nothing, resets to root
            continue
        table[value] = code
    if terminator is None:
        raise SystemExit('dictionary has no empty-value (terminator) entry -- refusing to guess one')
    return table, terminator


def build_tree(table, terminator):
    """The decoder's view: a binary tree of bits -> value."""
    root = {}
    for value, code in list(table.items()) + [('', terminator)]:
        node = root
        for b in code:
            node = node.setdefault(b, {})
        if 'v' in node and node['v'] != value:
            raise SystemExit('dictionary is ambiguous: %r and %r share a code' % (node['v'], value))
        node['v'] = value
    return root


def encode(text, table, terminator):
    """Longest-match-first, because the dictionary holds phrases as well as characters and a
    character-only encoding would still decode correctly but be needlessly long."""
    by_len = sorted({len(k) for k in table}, reverse=True)
    bits, i = [], 0
    while i < len(text):
        for n in by_len:
            chunk = text[i:i + n]
            if chunk in table:
                bits.append(table[chunk])
                i += n
                break
        else:
            raise SystemExit('no code for %r at %d -- the dictionary cannot express this string' % (text[i], i))
    bits.append(terminator)
    return ''.join(bits)


def pack(bits):
    """Bits -> bytes, with the format's first-byte quirk: byte 0 carries only six bits."""
    first, rest = bits[:6], bits[6:]
    out = [int(first.ljust(6, '0'), 2) & 0x3F]      # top two bits left clear
    for i in range(0, len(rest), 8):
        out.append(int(rest[i:i + 8].ljust(8, '0'), 2))
    return bytes(out)


def unpack(data):
    """Bytes -> bits, mirroring the decoder: first byte from mask 0x20, the rest from 0x80."""
    bits = []
    for i, byte in enumerate(data):
        mask = 0x20 if i == 0 else 0x80
        while mask:
            bits.append('1' if byte & mask else '0')
            mask >>= 1
    return ''.join(bits)


def decode(data, tree):
    """Exactly what the box's decoder does, so a round trip proves the encoder."""
    out, node = [], tree
    for bit in unpack(data):
        node = node.get(bit)
        if node is None:
            return ''.join(out), False       # ran off the tree: the encoder is wrong
        if 'v' in node:
            if node['v'] == '':
                return ''.join(out), True    # terminator
            out.append(node['v'])
            node = tree
    return ''.join(out), False               # ran out of bits without terminating


def encode_title(text, dict_path=DEFAULT_DICT):
    table, terminator = read_dictionary(dict_path)
    return pack(encode(text, table, terminator))


if __name__ == '__main__':
    path = sys.argv[1] if len(sys.argv) > 1 else DEFAULT_DICT
    table, terminator = read_dictionary(path)
    tree = build_tree(table, terminator)
    print('dictionary %s: %d entries, terminator %r' % (path, len(table), terminator))

    # THE ROUND TRIP IS THE WHOLE POINT. Titles from the era, plus the cases most likely to break an
    # encoder: a space-led string, punctuation, digits, and a phrase that should match a multi-char
    # dictionary entry rather than being spelled out.
    samples = ['The Simpsons', 'Sky News', 'Football', 'News at Ten',
               'The X-Files', 'Coronation Street', 'A', ' leading space', 'Digits 1998']
    bad = 0
    for s in samples:
        try:
            packed = pack(encode(s, table, terminator))
        except SystemExit as e:
            print('  ENCODE FAILED %-20r %s' % (s, e)); bad += 1; continue
        got, terminated = decode(packed, tree)
        ok = (got == s and terminated)
        if not ok: bad += 1
        print('  %-20r -> %2d bytes -> %-22r %s' % (s, len(packed), got, 'ok' if ok else 'MISMATCH'))
    print('\n%d of %d round-tripped' % (len(samples) - bad, len(samples)))
    sys.exit(1 if bad else 0)

File: internal/broadcast/huffman.go
package broadcast

import (
	"bufio"
	"fmt"
	"os"
	"sort"
	"strings"
)

// The Sky/OpenTV EPG carries programme text Huffman-coded against a dictionary
// the box already holds in flash. Every public implementation of this format
// READS the stream; a listings feed has to WRITE one, which means encoding —
// and an encoder has a property a decoder does not, in that it can be checked
// against the decoder it is meant to feed. So both halves live here and the
// round trip is the test.
//
// The dictionary itself is not in this repository. It is third-party data of
// uncertain provenance and is handled like the firmware: local, gitignored,
// and described in dictionaries/MANIFEST.md.

// HuffmanDictionary is one loaded Sky EPG code table.
type HuffmanDictionary struct {
	codes      map[string]string // value -> bit string
	terminator string            // the empty-value node: emits nothing, ends the text
	lengths    []int             // distinct value lengths, longest first
}

// LoadHuffmanDictionary reads the `<value>=<bits>` table.
//
// The parse mirrors openTVtoXML's huffman_read_dictionary(), which tries three
// scanf patterns IN ORDER: a single character, then a multi-character phrase,
// then an empty value. That order is load-bearing rather than incidental —
// splitting naively on '=' mis-reads both the line that maps SPACE (it begins
// with one) and any phrase containing an '='. Partitioning on the FIRST '='
// reproduces all three cases.
func LoadHuffmanDictionary(path string) (*HuffmanDictionary, error) {
	file, err := os.Open(path) // #nosec G304 -- operator-supplied dictionary, like the firmware
	if err != nil {
		return nil, fmt.Errorf("broadcast: huffman dictionary: %w", err)
	}
	// Read-only, so a close error tells us nothing we can act on, but it is
	// acknowledged rather than dropped.
	defer func() { _ = file.Close() }()

	dict := &HuffmanDictionary{codes: map[string]string{}}
	seen := map[int]bool{}
	scanner := bufio.NewScanner(file)
	for scanner.Scan() {
		line := strings.TrimRight(scanner.Text(), "\r")
		value, bits, found := strings.Cut(line, "=")
		if !found || bits == "" || strings.Trim(bits, "01") != "" {
			continue
		}
		if value == "" {
			dict.terminator = bits
			continue
		}
		dict.codes[value] = bits
		if !seen[len(value)] {
			seen[len(value)] = true
			dict.lengths = append(dict.lengths, len(value))
		}
	}
	if err := scanner.Err(); err != nil {
		return nil, fmt.Errorf("broadcast: huffman dictionary: %w", err)
	}
	if dict.terminator == "" {
		// Without it an encoder cannot say where the text stops, and the box
		// would read whatever followed as more programme title.
		return nil, fmt.Errorf("broadcast: huffman dictionary %s has no terminator entry; refusing to guess one", path)
	}
	if len(dict.codes) == 0 {
		return nil, fmt.Errorf("broadcast: huffman dictionary %s is empty", path)
	}
	sort.Sort(sort.Reverse(sort.IntSlice(dict.lengths)))
	return dict, nil
}

// Entries reports how many values the dictionary can code, so a caller can say
// which table it loaded rather than assuming.
func (d *HuffmanDictionary) Entries() int { return len(d.codes) }

// Encode packs text the way the box's decoder unpacks it.
//
// Longest match first: the dictionary holds phrases as well as characters, and
// a character-only encoding would still decode correctly but be needlessly
// long — which matters because a title section has a 12-bit length and a fixed
// number of records to fit inside it.
func (d *HuffmanDictionary) Encode(text string) ([]byte, error) {
	var bits strings.Builder
	for i := 0; i < len(text); {
		matched := false
		for _, n := range d.lengths {
			if i+n > len(text) {
				continue
			}
			if code, ok := d.codes[text[i:i+n]]; ok {
				bits.WriteString(code)
				i += n
				matched = true
				break
			}
		}
		if !matched {
			// Silently dropping it would broadcast a title that is not the one
			// asked for, which is the failure this whole package is careful
			// about elsewhere.
			return nil, fmt.Errorf("broadcast: no huffman code for %q at %d in %q", text[i], i, text)
		}
	}
	bits.WriteString(d.terminator)
	return packHuffman(bits.String()), nil
}

// packHuffman turns a bit string into the bytes the decoder expects.
//
// THE FIRST BYTE CARRIES ONLY SIX BITS. The decoder starts byte 0 at mask 0x20
// and every later byte at 0x80, so byte 0's top two bits are skipped entirely.
// It is the format, not a bug in the reference reader, and an encoder that
// packs eight bits into byte 0 produces plausible-looking output that decodes
// to nonsense — which is exactly the kind of failure this box gives no error
// for.
func packHuffman(bits string) []byte {
	out := []byte{bitsToByte(bits, 0, 6)}
	for i := 6; i < len(bits); i += 8 {
		out = append(out, bitsToByte(bits, i, 8))
	}
	return out
}

func bitsToByte(bits string, at, width int) byte {
	var value byte
	for i := 0; i < width; i++ {
		value <<= 1
		if at+i < len(bits) && bits[at+i] == '1' {
			value |= 1
		}
	}
	return value
}

// Decode is the box's decoder, reimplemented so the encoder can be proved
// against it rather than trusted. It reports whether the text terminated
// properly: a stream that runs off the tree or out of bits is an encoder bug,
// and saying "here is the text so far" without saying that would hide it.
func (d *HuffmanDictionary) Decode(data []byte) (string, bool) {
	type node struct {
		children map[byte]*node
		value    string
		isLeaf   bool
	}
	root := &node{children: map[byte]*node{}}
	insert := func(code, value string) {
		cur := root
		for i := 0; i < len(code); i++ {
			next, ok := cur.children[code[i]]
			if !ok {
				next = &node{children: map[byte]*node{}}
				cur.children[code[i]] = next
			}
			cur = next
		}
		cur.value, cur.isLeaf = value, true
	}
	for value, code := range d.codes {
		insert(code, value)
	}
	insert(d.terminator, "")

	var out strings.Builder
	cur := root
	for i, b := range data {
		mask := byte(0x80)
		if i == 0 {
			mask = 0x20
		}
		for ; mask != 0; mask >>= 1 {
			bit := byte('0')
			if b&mask != 0 {
				bit = '1'
			}
			next, ok := cur.children[bit]
			if !ok {
				return out.String(), false // ran off the tree
			}
			cur = next
			if cur.isLeaf {
				if cur.value == "" {
					return out.String(), true // the terminator
				}
				out.WriteString(cur.value)
				cur = root
			}
		}
	}
	return out.String(), false // ran out of bits without terminating
}
